core/invoker/src/main/scala/whisk/core/containerpool/docker/ProcessRunner.scala

```diff
@@ -52,17 +52,75 @@ trait ProcessRunner {
-      case (0, stdout, _, scheduled) =>
+      case (ExitStatus(0), stdout, _, scheduled) =>
         scheduled.foreach(_.cancel())
         Future.successful(stdout)
-      case (code, stdout, stderr, scheduled) =>
+      case (exitStatus, stdout, stderr, scheduled) =>
         scheduled.foreach(_.cancel())
-        Future.failed(ProcessRunningException(code, stdout, stderr))
+        timeout match {
+          case t: FiniteDuration if exitStatus.terminatedBySIGTERM =>
+            Future.failed(ProcessTimeoutException(timeout, exitStatus, stdout, stderr))
+          case _ => Future.failed(ProcessRunningException(exitStatus, stdout, stderr))
+        }
     }
+}
+
+case class ExitStatus(statusValue: Int) {
+
+  // Based on The Open Group Base Specifications Issue 7, 2018 edition:
+  // Shell & Utilities - Shell Command Language - 2.8.2 Exit Status for Commands
+  // http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_08_02
+  val STATUS_SUCCESSFULL = 0
+  val STATUS_NOT_EXECUTABLE = 126
+  val STATUS_NOT_FOUND = 127
+  // When a command is stopped by a signal, the exit status is 128 + signal numer
+  val STATUS_SIGNAL = 128
 
+  // Based on The Open Group Base Specifications Issue 7, 2018 edition:
+  // Shell & Utilities - Utilities - kill
+  // http://pubs.opengroup.org/onlinepubs/9699919799/utilities/kill.html
+  val SIGHUP = 1
+  val SIGINT = 2
+  val SIGQUIT = 3
+  val SIGABRT = 6
+  val SIGKILL = 9
+  val SIGALRM = 14
+  val SIGTERM = 15
+
+  override def toString(): String = {
+    def signalAsString(signal: Int): String = {
+      signal match {
+        case SIGHUP  => "SIGHUP"
+        case SIGINT  => "SIGINT"
+        case SIGQUIT => "SIGQUIT"
+        case SIGABRT => "SIGABRT"
+        case SIGKILL => "SIGKILL"
+        case SIGALRM => "SIGALRM"
+        case SIGTERM => "SIGTERM"
+        case _       => signal.toString
+      }
+    }
+
+    val detail = statusValue match {
+      case STATUS_SUCCESSFULL                => "successful"
+      case STATUS_NOT_EXECUTABLE             => "not executable"
+      case STATUS_NOT_FOUND                  => "not found"
+      case _ if statusValue >= STATUS_SIGNAL => "terminated by signal " + signalAsString(statusValue - STATUS_SIGNAL)
+      case _                                 => "unsuccessful"
+    }
+
+    s"$statusValue ($detail)"
+  }
+
+  val successful = statusValue == STATUS_SUCCESSFULL
+  val terminatedBySIGTERM = (statusValue - STATUS_SIGNAL) == SIGTERM
 }
 
-case class ProcessRunningException(exitCode: Int, stdout: String, stderr: String)
-    extends Exception(s"code: $exitCode ${if (exitCode == 143) "(killed)" else ""}, stdout: $stdout, stderr: $stderr")
+case class ProcessRunningException(exitStatus: ExitStatus, stdout: String, stderr: String)
+    extends Exception(s"code: $exitStatus, stdout: $stdout, stderr: $stderr")
+
+case class ProcessTimeoutException(timeout: Duration, exitStatus: ExitStatus, stdout: String, stderr: String)
```

Comment:
It may be better to have `ProcessTimeoutException` extend `ProcessRunningException` such that any future code which wants to capture any process related exception (including timeout) can catch it in a generic manner

Reply:
Good advice. Done.